### PR TITLE
Support invoking LaunchCudaKernel

### PR DESCRIPTION
**Pull request by changhong94**

### CODE REVIEW

File: cudaproxy/src/cudaproxy.rs [hunk 1]
```diff
@@ -41,16 +40,93 @@ pub extern "C" fn cudaDeviceSynchronize() -> usize {
     };
 }
 
+#[no_mangle]
+pub extern "C" fn __cudaRegisterFatBinary(fatCubin: &FatHeader) -> *mut u64 {
+    println!("Hijacked __cudaRegisterFatBinary(fatCubin:{:#x?})", fatCubin);
+    let addrFatCubin = fatCubin.text as *const _ as *const u64;
+    println!("hochan ProxyCommand::CudaRegisterFatBinary: {:x}, fatCubin.text.header_size:{:x}, fatCubin.text.size: {:x}, addrFatCubin: {:x}",
```
Comment (changhong94) [inline cudaproxy/src/cudaproxy.rs:47]: All the logging with word "hochan" will be removed before merge.

File: config.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "DebugLevel"    : "Error",
```
Comment (changhong94) [inline config.json:2]: Revert

File: cudaproxy/src/cudaproxy.rs [hunk 3]
```diff
@@ -41,16 +40,93 @@ pub extern "C" fn cudaDeviceSynchronize() -> usize {
...
+    // let bytes = std::slice::from_raw_parts(fatCubin.text as *const _ as u64 as *const u8, len);
+    println!("hochan fatCubin.text addr {:x}", fatCubin.text as *const _ as u64);
+    // println!("hochan !!!0 {:x?}", bytes);
+    println!("hochan addr {:x?}, addrFatCubin {:x?}",addr,addrFatCubin);
+
+    let result = 0 as *mut u64;
+    unsafe {
+        syscall4(SYS_PROXY, ProxyCommand::CudaRegisterFatBinary as usize, len, fatCubin.text as *const _ as usize, result as usize);
+    }
+    return result;
```
Comment (changhong94) [inline cudaproxy/src/cudaproxy.rs:60]: Check result

Comment (changhong94) [reply cudaproxy/src/cudaproxy.rs:60]: Check how to get correct result.

File: qvisor/src/vmspace/nvidia.rs [hunk 4]
```diff
@@ -56,16 +71,90 @@ pub fn  NvidiaProxy(cmd: ProxyCommand, parameters: &ProxyParameters) -> Result<i
...
             return Ok(ret as i64);
         }
         ProxyCommand::CudaMemcpy => {
             return CudaMemcpy(handler, parameters);
         }
+        ProxyCommand::CudaRegisterFatBinary => {
+            let fatElfHeader = unsafe { &*(parameters.para2 as *const u8 as *const FatElfHeader) };
+            let moduleKey = parameters.para3;
+            error!("hochan moduleKey:{:x}", moduleKey);
+            match GetFatbinInfo(parameters.para2, *fatElfHeader){
```
Comment (changhong94) [inline qvisor/src/vmspace/nvidia.rs:100]: Copy happens here. Remove *. Check other places with similar issue.

File: qvisor/src/vmspace/nvidia.rs [hunk 5]
```diff
@@ -56,16 +71,90 @@ pub fn  NvidiaProxy(cmd: ProxyCommand, parameters: &ProxyParameters) -> Result<i
...
+            error!("hochan CudaMalloc after parameters:{:x?} ret {:x?}", parameters, ret);
             return Ok(ret as i64);
         }
         ProxyCommand::CudaMemcpy => {
             return CudaMemcpy(handler, parameters);
         }
+        ProxyCommand::CudaRegisterFatBinary => {
+            let fatElfHeader = unsafe { &*(parameters.para2 as *const u8 as *const FatElfHeader) };
+            let moduleKey = parameters.para3;
+            error!("hochan moduleKey:{:x}", moduleKey);
```
Comment (changhong94) [inline qvisor/src/vmspace/nvidia.rs:99]: Check moduleKey when set value. What's its usage.

File: qvisor/src/vmspace/nvidia.rs [hunk 6]
```diff
@@ -56,16 +71,90 @@ pub fn  NvidiaProxy(cmd: ProxyCommand, parameters: &ProxyParameters) -> Result<i
...
+
+                let mut hfunc:u64 = 0;
+                let ret = unsafe { cuda_driver_sys::cuModuleGetFunction(
+                    &mut hfunc as *mut _ as u64 as *mut CUfunction, 
+                    *(&mut module as *mut _ as u64 as *mut CUmodule), 
+                    CString::new(deviceName).unwrap().clone().as_ptr()) };
+                FUNCTIONS.lock().insert(info.hostFun, hfunc);
+                error!("hochan cuModuleGetFunction ret {:x?}, hfunc {:x}, &hfunc {:x}, FUNCTIONS  {:x?}", ret, hfunc, &hfunc, FUNCTIONS.lock());
+
+                let kernelInfo = KERNEL_INFOS.lock().get(&deviceName.to_string()).unwrap().clone();
```
Comment (changhong94) [inline qvisor/src/vmspace/nvidia.rs:129]: Change unwrap to match->return err. In this way, no hung in qvisor.

**Comment by shrik3** [comment]
Hi, I'm not a reviewer of your works, but since you are merging into main, I would ask you to kindly take some effort to rebase the commit history, here are some suggestions:

0. IMHO this PR should not need more than 2 commits
1. obviously, fix the conflicts
2. [59cdac3](https://github.com/QuarkContainer/Quark/pull/1054/commits/59cdac31e300a770600fb472506d2177bfd3af59) makes no sense in this PR, it should be the other way around.
3. there are consecutive commits named "Refine Code". Please squash them into one.
4. commit message should be meaningful, for example "temp" isn't a good one.

Please don't take this as criticism. I just want to point out the bad practice.

cheers.

File: qvisor/src/vmspace/nvidia.rs (hunk 5 above)
Comment (changhong94) [reply qvisor/src/vmspace/nvidia.rs:99]: Module key is generated in cudaproxy.rs, then be as the key of the fatbin.

File: qvisor/src/vmspace/nvidia.rs (hunk 4 above)
Comment (changhong94) [reply qvisor/src/vmspace/nvidia.rs]: Fixed

**Pull request closed without merging**